Python语言基础

本章是python语言的基础部分，也是后续内容的基础。

1 Python数据类型

1.1 字符串

在Python中用引号引起来的字符集称之为字符串，比如：'hello'、"my Python"、"2+3"等都是字符串 Python中字符串中使用的引号可以是单引号、双引号跟三引号

In [913]:
print ('hello world!')

hello world!


In [914]:
c = 'It is a "dog"!'
print (c)

It is a "dog"!


In [915]:
c1= "It's a dog!"
print (c1)

It's a dog!


In [916]:
c2 = """hello
world
!"""
print (c2)

hello
world
!


* 转义字符'\'

转义字符\可以转义很多字符，比如\n表示换行，\t表示制表符，字符\本身也要转义，所以\表示的字符就是\

In [917]:
print ('It\'s a dog!')
print ("hello world!\nhello Python!")
print ('\\\t\\')

It's a dog!
hello world!
hello Python!
\	\


原样输出引号内字符串可以使用在引号前加r

In [918]:
print (r'\\\t\\')

\\\t\\


* 子字符串及运算


In [919]:
s = 'Python'
print( 'Py' in s)

print( 'py' in s)

True
False


取子字符串有两种方法，使用[]索引或者切片运算法[:]，这两个方法使用面非常广


In [920]:
print (s[2])

t


In [921]:
print (s[1:4])

yth


* 字符串连接与格式化输出

In [922]:
word1 = '"hello"'
word2 = '"world"'
sentence = word1.strip('"') + ' ' + word2.strip('"') + '!'

print( 'The first word is %s, and the second word is %s' %(word1, word2))
print (sentence)

The first word is "hello", and the second word is "world"
hello world!


1.2 整数与浮点数

整数

Python可以处理任意大小的整数，当然包括负整数，在程序中的表示方法和数学上的写法一模一样

In [923]:
i = 7
print (i)

7


In [924]:
7 + 3

10

In [925]:
7 - 3

4

In [926]:
7*3

21

In [927]:
7**3

343

In [928]:
7/3  #Python3之后，整数除法和浮点数除法已经没有差异

2.3333333333333335

In [929]:
7%3

1

In [930]:
7//3

2

浮点数

In [931]:
7.0 / 3

2.3333333333333335

In [932]:
3.14 * 10 ** 2

314.0

其它表示方法

In [933]:
0b1111

15

In [934]:
0xff

255

In [935]:
1.2e-5

1.2e-05

更多运算

In [936]:
import math

print (math.log(math.e)) # 更多运算可查阅文档

1.0


1.3 布尔值

In [937]:
True

True

In [938]:
False

False

In [939]:
True and False

False

In [940]:
True or False

True

In [941]:
not True

False

In [942]:
True + False

1

In [943]:
18 >= 6 * 3 or 'py' in 'Python'

True

In [944]:
18 >= 6 * 3 and 'py' in 'Python'

False

In [945]:
18 >= 6 * 3 and 'Py' in 'Python'

True

1.4 日期时间

In [946]:
import time

now = time.strptime('2016-07-20', '%Y-%m-%d')
print (now)

time.struct_time(tm_year=2016, tm_mon=7, tm_mday=20, tm_hour=0, tm_min=0, tm_sec=0, tm_wday=2, tm_yday=202, tm_isdst=-1)


In [947]:
time.strftime('%Y-%m-%d', now)

'2016-07-20'

In [948]:
import datetime

someDay = datetime.date(1999,2,10)
anotherDay = datetime.date(1999,2,15)
deltaDay = anotherDay - someDay
deltaDay.days

5

还有其他一些datetime格式

![](images/datetime_format.jpg)

* 查看变量类型

In [949]:
type(None)

NoneType

In [950]:
type(1.0)

float

In [951]:
type(True)

bool

In [952]:
s="NoneType"
type(s)

str

* 类型转换

In [953]:
str(10086)

'10086'

In [954]:
float(10086)

10086.0

In [955]:
int('10086')

10086

In [956]:
complex(10086)

(10086+0j)

2 Python数据结构

列表（list）、元组（tuple）、集合（set）、字典（dict）

2.1 列表(list)

用来存储一连串元素的容器，列表用[]来表示，其中元素的类型可不相同。

In [957]:
mylist= [0, 1, 2, 3, 4, 5]
print (mylist)

[0, 1, 2, 3, 4, 5]


列表索引和切片

In [958]:
# 索引从0开始，含左不含右
print ('[4]=', mylist[4])
print ('[-4]=', mylist[-4])
print ('[0:4]=', mylist[0:4])
print ('[:4]=', mylist[:4])#dddd
print( '[4:]=', mylist[4:])
print ('[0:4:2]=', mylist[0:4:2])
print ('[-5:-1:]=', mylist[-5:-1:])
print ('[-2::-1]=', mylist[-2::-1])

[4]= 4
[-4]= 2
[0:4]= [0, 1, 2, 3]
[:4]= [0, 1, 2, 3]
[4:]= [4, 5]
[0:4:2]= [0, 2]
[-5:-1:]= [1, 2, 3, 4]
[-2::-1]= [4, 3, 2, 1, 0]


修改列表

In [959]:
mylist[3] = "小月"
print (mylist[3])

mylist[5]="小楠"
print (mylist[5])

mylist[5]=19978
print (mylist[5])

小月
小楠
19978


In [960]:
print (mylist)

[0, 1, 2, '小月', 4, 19978]


插入元素

In [961]:
mylist.append('han') # 添加到尾部
mylist.extend(['long', 'wan'])
print (mylist)

[0, 1, 2, '小月', 4, 19978, 'han', 'long', 'wan']


In [962]:
scores = [90, 80, 75, 66]
mylist.insert(1, scores) # 添加到指定位置
mylist

[0, [90, 80, 75, 66], 1, 2, '小月', 4, 19978, 'han', 'long', 'wan']

删除元素

In [963]:
print (mylist.pop(1)) # 该函数返回被弹出的元素，不传入参数则删除最后一个元素
print (mylist)

[90, 80, 75, 66]
[0, 1, 2, '小月', 4, 19978, 'han', 'long', 'wan']


判断元素是否在列表中等

In [964]:
print( 'wan' in mylist)
print ('han' not in mylist)

True
False


In [965]:
mylist.count('wan')

1

In [966]:
mylist.index('wan')

8

range函数生成整数列表

In [967]:
X = []
for x in range(10):
    X.append(x)
print(X)
X = []
for x in range(-5, 5):
    X.append(x)
print(X)
X = []
for x in range(-10, 10, 2):
    X.append(x)
print(X)
X = []
for x in range(16, 10, -1):
    X.append(x)
print(X)

[0, 1, 2, 3, 4, 5, 6, 7, 8, 9]
[-5, -4, -3, -2, -1, 0, 1, 2, 3, 4]
[-10, -8, -6, -4, -2, 0, 2, 4, 6, 8]
[16, 15, 14, 13, 12, 11]


2.2 元组(tuple)
元组类似列表，元组里面的元素也是进行索引计算。列表里面的元素的值可以修改，而元组里面的元素的值不能修改，只能读取。
元组的符号是()。

In [968]:
studentsTuple = ("ming", "jun", "qiang", "wu", scores)
studentsTuple

('ming', 'jun', 'qiang', 'wu', [90, 80, 75, 66])

In [969]:
try:
    studentsTuple[1] = 'fu'
except TypeError:
    print ('TypeError')

TypeError


In [970]:
scores[1]= 100
studentsTuple

('ming', 'jun', 'qiang', 'wu', [90, 100, 75, 66])

In [971]:
'ming' in studentsTuple

True

In [972]:
studentsTuple[0:4]

('ming', 'jun', 'qiang', 'wu')

In [973]:
studentsTuple.count('ming')

1

In [974]:
studentsTuple.index('jun')

1

In [975]:
len(studentsTuple)

5

2.3 集合(set)

Python中集合主要有两个功能，一个功能是进行集合操作，另一个功能是消除重复元素。
集合的格式是：set()，其中()内可以是列表、字典或字符串，因为字符串是以列表的形式存储的。